aten/src/ATen/native/cuda/Normalization.cu

```diff
@@ -0,0 +1,444 @@
+  }
+
+  AT_CHECK(in_dim < dim || t.is_contiguous(), "need contiguous or <= 3d tensor");
+  std::vector<int64_t> sizes(dim);
+  std::vector<int64_t> strides(dim);
+  for (int i = 0; i < in_dim || i < dim; ++i) {
+    if (i < dim && i < in_dim) {
+      sizes[i] = t.size(i);
+      strides[i] = t.stride(i);
+    } else if (i < dim) {
+      sizes[i] = 1;
+      strides[i] = 0;
+    } else {
+      sizes[dim - 1] *= t.size(i);
+      strides[dim -1] = 1;
+    }
+  }
+  // evil trick to get adjusted 2d tensors to have large dimension last
+  if (dim == 3 && sizes[0] > sizes[2]) {
+    std::swap(sizes[0], sizes[2]);
+    std::swap(strides[0], strides[2]);
+  }
+  return PackedTensorAccessor<scalar_t, dim, at::RestrictPtrTraits>(t.data<scalar_t>(), sizes.data(), strides.data());
+}
+
+template<typename scalar_t>
+std::tuple<Tensor, Tensor, Tensor> batch_norm_cuda_template(const Tensor& input_, const Tensor& weight_, const Tensor& bias_,
+							    const Tensor& running_mean_, const Tensor& running_var_,
+							    bool train, double momentum, double epsilon) {
+
+  using accscalar_t = at::acc_type<scalar_t, true>;
+  Tensor output_= at::empty_like(input_);
+  int64_t n_input = input_.size(1);
+  Tensor save_mean_;
+  Tensor save_invstd_;
+  auto input_options = input_.options();
+  if (input_options.dtype() == ScalarType::Half) {
+    input_options.dtype(ScalarType::Float);
+  }
+  if (train) {
+    save_mean_ = at::empty({n_input}, input_options);
+    save_invstd_ = at::empty({n_input}, input_options);
+  } else {
+    save_mean_ = at::empty({0}, input_options);
+    save_invstd_ = at::empty({0}, input_options);
+  }
+  auto input = reshaped_packed_accessor<scalar_t, 3>(input_);
+  auto output = reshaped_packed_accessor<scalar_t, 3>(output_);
+  auto weight = reshaped_packed_accessor<scalar_t, 1>(weight_);
+  auto bias = reshaped_packed_accessor<scalar_t, 1>(bias_);
+  auto running_mean = reshaped_packed_accessor<scalar_t, 1>(running_mean_);
+  auto running_var = reshaped_packed_accessor<scalar_t, 1>(running_var_);
+  auto save_mean = reshaped_packed_accessor<accscalar_t, 1>(save_mean_);
+  auto save_invstd = reshaped_packed_accessor<accscalar_t, 1>(save_invstd_);
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  constexpr int max_blocks_per_input = 60000;
+  int feature_blocks = std::min<int>(input.size(2), max_blocks_per_input);
+  int batch_blocks   = std::min<int>(input.size(0), max_blocks_per_input / feature_blocks);
+  dim3 blocks(batch_blocks, (input.size(1)+127)/128, feature_blocks);
+  dim3 threads(1, 128);
+  if (!train) {
+    batch_norm_transform_input_kernel<scalar_t, accscalar_t, false> <<<blocks, threads, 0, stream>>>
+      (input, output, running_mean, running_var, weight, bias, epsilon);
+  } else {
+    dim3 blocks_red(input.size(1));
+    dim3 threads_red(getNumThreads(input.size(2)));
+    batch_norm_collect_statistics_kernel<scalar_t, accscalar_t> <<<blocks_red, threads_red, 0, stream>>>
```

Comment:
This is the part where stats collection and transform use different thread/block allocation, right?

Reply:
Right.